Fix Application#trace which would interleave CR in a multi-threaded context

michaeljbishop opened the pull request:
Running rake with tracing on and executing multitasks would show
a trace with multiple traces on a single line and other lines having
multiple CRs. This is because `#trace` used `IO#puts` behind the scenes.

`IO#puts` first calls `IO#write` with the string, followed by another `IO#write`
with the CR. Because of the gap in-between calls, another thread
could sneak in a CR or another string.

The fix is to make sure calls to `Application#trace` create a single string with
a CR appended before calling `IO#print`. This will output a single
string with a trailing CR in thread-safe way.

Example output showing the error:

```
** Execute 98** Execute 68** Execute 69

** Execute 74

** Execute 78** Execute 81
** Execute 84** Execute 100

** Execute 89
```

michaeljbishop committed:
Fix Application#trace which would interleave CR in a multi-threaded context.

Running rake with tracing on and executing multitasks would show
a trace with multiple traces on a single line and other lines having
multiple CRs. This is because #trace used IO#puts behind the scenes.

IO#puts first calls #write with the string, followed by another #write
with the CR. Because of the gap in-between calls, another thread
could sneak in a CR or another string.

The fix is to make sure calls to trace create a single string with
a CR appended before calling IO#print. This will output a single
string with a trailing CR in thread-safe way.

Example output showing the error:

** Execute 98** Execute 68** Execute 69

** Execute 74

** Execute 78** Execute 81
** Execute 84** Execute 100

** Execute 89

jimweirich commented:
Did some refactoring of this request.  First, a test that relies on random sleeps in tasks is a red flag to me, so I added tests that insured that trace called print exactly once and verified that the output was as expected.  Then I dropped the task-heavy test, and finally moved the trace logic (and related tests) into its own module

michaeljbishop commented:
Thanks for doing that!

BTW, I'm flummoxed by an issue I saw in MultiTask vs Task and I was wondering if you might have the answer. I posted to Rake-Devel @

  "Why do MultiTask and Task send different args to their prerequisites?"

  http://rubyforge.org/pipermail/rake-devel/2012-November/000890.html

If you have a moment, can you look at that? I'm reluctant to make them consistent since I looked through the history and they've invoked inconsistently for a while. I just wanted to make sure it wasn't intentional.

I've made a fix privately for this but didn't make a pull request since I wasn't so sure of it.

  https://github.com/michaeljbishop/rake/commit/ca283160b610c95e1f28cf561af8811fd37d10c3

On Nov 21, 2012, at 11:51 AM, Jim Weirich notifications@github.com wrote:

> Did some refactoring of this request. First, a test that relies on random sleeps in tasks is a red flag to me, so I added tests that insured that trace called print exactly once and verified that the output was as expected. Then I dropped the task-heavy test, and finally moved the trace logic (and related tests) into its own module
> 
> —
> Reply to this email directly or view it on GitHub.
